### PR TITLE
resolveTextDirection using Character.getDirectionality (fixes #90)

### LINKED ISSUE #90: [win] slow org.eclipse.swt.internal.BidiUtil.resolveTextDirection()
When i open my eclipse commiter workspace it takes some seconds till everything is shown and i can finally search for a type (ctrl shift T).
I sampled a bit and (beside that the files are refreshed - i understand that need) the longest time (6sec) is spend in 
org.eclipse.swt.internal.win32.OS.GetCharacterPlacement[native] ()
inside BidiUtil.resolveTextDirection().

I don't even need any Bidi support so that feels just bad. Is there a way to turn bidi support of or a chance of less costly detection of bidi characters (without native calls **for every single character**)?
image

according to the source that bidi feature was added in 2015 https://bugs.eclipse.org/bugs/show_bug.cgi?id=467463
but is only recently used in this context since 2022-04  https://bugs.eclipse.org/bugs/show_bug.cgi?id=579626
@niraj-modi

### PR DESCRIPTION
**Pull request by jukzi**
prevent slow systemcalls for every character

### CODE REVIEW

**Review by niraj-modi — CHANGES_REQUESTED** [review]

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/internal/BidiUtil.java [hunk 1]
```diff
@@ -470,51 +470,23 @@ public static void removeLanguageListener (Control control) {
...
+		char c = text.charAt(i);
+		byte directionality = Character.getDirectionality(c);
+		// XXX what is a "strong bidirectional character"?
+		// impl taken from org.eclipse.jface.util.BaseTextDirectionSegmentListener.isRTLValue(String)
+		if (directionality == Character.DIRECTIONALITY_RIGHT_TO_LEFT
+				|| directionality == Character.DIRECTIONALITY_RIGHT_TO_LEFT_ARABIC
+				|| directionality == Character.DIRECTIONALITY_ARABIC_NUMBER)
+			return SWT.RIGHT_TO_LEFT;
+		else if (directionality == Character.DIRECTIONALITY_LEFT_TO_RIGHT) {
+			return SWT.LEFT_TO_RIGHT;
```
[the remark this answers is not in the document]
Comment (niraj-modi) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/internal/BidiUtil.java:486]: Try below relevant scenarios for testing MIX of English and RTL in single line:
1. https://github.com/eclipse-platform/eclipse.platform.swt/blob/master/tests/org.eclipse.swt.tests.win32/ManualTests/org/eclipse/swt/tests/win32/snippets/Bug579626_English_ArabicTest.java
2. Opening above file source-code in Eclipse self-hosted mode.
3. Output of above file on console in Eclipse self-hosted mode.

[the remark this answers is not in the document]
Comment (jukzi) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/internal/BidiUtil.java:486]: @niraj-modi it sounds like you know more about Bidi then me. Could you please take over and use my commit as an rough draft on how we could improve performance?
I will also don't have time nor hardware to adapt/test it n other hardware.

**Comment by jukzi** [comment]
@niraj-modi do you plan to work on it or otherwise review my correction?

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/internal/BidiUtil.java [hunk 2]
```diff
@@ -456,66 +456,78 @@ public static void removeLanguageListener (long hwnd) {
...
+		if (textDirection == SWT.RIGHT_TO_LEFT) {
+			break;
 		}
 	}
-	OS.ReleaseDC (0, hdc);
-	OS.HeapFree(hHeap, 0, lpOrder);
 	return textDirection;
+}
+
+public static int getStrongDirection(byte directionality) {
```
Comment (niraj-modi) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/internal/BidiUtil.java:490]: This method should be non-public, default access should be fine:
static int getStrongDirection(byte directionality)

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/internal/BidiUtil.java [hunk 3]
```diff
@@ -456,66 +456,78 @@ public static void removeLanguageListener (long hwnd) {
...
+	switch (directionality) {
+	// Strong:
+	case Character.DIRECTIONALITY_LEFT_TO_RIGHT:
+		return SWT.LEFT_TO_RIGHT;// Strong bidirectional character type "L" in the Unicode specification.
+	case Character.DIRECTIONALITY_RIGHT_TO_LEFT:
+		return SWT.RIGHT_TO_LEFT;// Strong bidirectional character type "R" in the Unicode specification.
+	case Character.DIRECTIONALITY_RIGHT_TO_LEFT_ARABIC:
+		return SWT.RIGHT_TO_LEFT;// Strong bidirectional character type "AL" in the Unicode specification.
```
Comment (niraj-modi) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/internal/BidiUtil.java:498]: Less code here, like below:
case Character.DIRECTIONALITY_RIGHT_TO_LEFT:
case Character.DIRECTIONALITY_RIGHT_TO_LEFT_ARABIC:
	return SWT.RIGHT_TO_LEFT;// Strong bidirectional character type "AL" in the Unicode specification.

**Review by niraj-modi — CHANGES_REQUESTED** [review]
Did a quick test with the Mix of English and RTL characters, it work fine.
Please see inline for code improvements.

**Comment by niraj-modi** [comment]
@[Phillipus](https://github.com/Phillipus)
Hi Phil,
Can you try/test this improvement on BidiUtil.resolveTextDirection which uses Java capability now ?

**Comment by niraj-modi** [comment]
> @niraj-modi do you plan to work on it or otherwise review my correction?

Thanks @jukzi for the revised PR, it's much improved now.
I have requested @[Phillipus](https://github.com/Phillipus) on help with further testing.

**Comment by jukzi** [comment]
regarding performance: this changes reduces org.eclipse.debug.tests.console.IOConsoleTests.testBug421303_StreamProcessingDeadlock() from ~ 5sec to 0.3sec
see also https://github.com/eclipse-platform/eclipse.platform.debug/issues/5

**Review by niraj-modi — APPROVED** [review]
Latest code changes, looks good to me.

**Comment by niraj-modi** [comment]
> @[Phillipus](https://github.com/Phillipus) Hi Phil, Can you try/test this improvement on BidiUtil.resolveTextDirection which uses Java capability now ?

We are still open to any feedback on this. Now changes can be directly tested in next Eclipse IBuild.

**Pull request merged**